src/webgpu/api/operation/rendering/draw.spec.ts

```diff
@@ -44,7 +28,330 @@ TODO:
+      .combine(poptions('count', [0, 6] as const))
+      .combine(poptions('first_instance', [0, 2] as const))
+      .combine(poptions('instance_count', [0, 4] as const))
+      .combine(pbool('indexed'))
+      .combine(pbool('indirect'))
+      .combine(poptions('vertex_buffer_offset', [0, 32] as const))
+      .expand(p => poptions('index_buffer_offset', p.indexed ? ([0, 16] as const) : [undefined]))
+      .expand(p => poptions('base_vertex', p.indexed ? ([0, 9] as const) : [undefined]))
+  )
+  .fn(t => {
+    const renderTargetSize = [36, 36];
+
+    // The test will split up the render target into a grid where triangles of
+    // increasing primitive id will be placed along the X axis, and triangles
+    // of increasing instance id will be placed along the Y axis. The size of the
+    // grid is based on the max primitive id and instance id used.
+    const numX = 3;
+    const numY = 6;
+    const tileSizeX = renderTargetSize[0] / numX;
+    const tileSizeY = renderTargetSize[1] / numY;
+
+    // |\
+    // |   \
+    // |______\
+    // Unit triangle shaped like this. 0-1 Y-down.
+    const bottomLeftTriangleVertices = /* prettier-ignore */ [
+      0.0, 0.0,
+      0.0, 1.0,
+      1.0, 1.0,
+    ];
+
+    // _______
+    // \      |
+    //    \   |
+    //       \|
+    // Unit triangle shaped like this. 0-1 Y-down.
+    const topRightTriangleVertices = /* prettier-ignore */ [
+      0.0, 0.0,
+      1.0, 1.0,
+      1.0, 0.0,
+    ];
+
+    const renderTarget = t.device.createTexture({
+      size: renderTargetSize,
+      usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC,
+      format: 'rgba8unorm',
+    });
+
+    const vertexModule = t.device.createShaderModule({
+      code: `
+[[builtin(vertex_index)]] var<in> vertex_index : u32;
+[[builtin(instance_index)]] var<in> instance_id : u32;
+[[location(0)]] var<in> vertexPosition : vec2<f32>;
+
+[[builtin(position)]] var<out> Position : vec4<f32>;
+[[stage(vertex)]] fn vert_main() -> void {
+  // 9u is the base_vertex offset. We select different vertices
+  // but apply the same x offset in the grid.
+  // 3u is the number of points in a triangle to convert from index
+  // to id.
+  var vertex_id : u32 = (vertex_index % 9u) / 3u;
+
+  var x : f32 = (vertexPosition.x + f32(vertex_id)) / ${numX}.0;
+  var y : f32 = (vertexPosition.y + f32(instance_id)) / ${numY}.0;
+
+  // (0,1) y-down space to (-1,1) y-up NDC
+  x = 2.0 * x - 1.0;
+  y = -2.0 * y + 1.0;
+  Position = vec4<f32>(x, y, 0.0, 1.0);
+}
+`,
+    });
+
+    const fragmentModule = t.device.createShaderModule({
+      code: `
+[[block]] struct Output {
+  [[offset(0)]] value : u32;
+};
+
+[[group(0), binding(0)]] var<storage> output : Output;
+
+[[location(0)]] var<out> fragColor : vec4<f32>;
+[[stage(fragment)]] fn frag_main() -> void {
+  output.value = 1u;
+  fragColor = vec4<f32>(0.0, 1.0, 0.0, 1.0);
+}
+`,
+    });
+
+    const pipeline = t.device.createRenderPipeline({
```

Comment:
nit: this could use the new render pipeline descriptor when webgpu/types is rolled to 0.0.44

Reply:
This was before that landed, but I'll update this now.

Reply:
Done.